functorch/experimental/_map.py

```diff
@@ -326,6 +345,24 @@ def map_fake_tensor_mode(f, num_mapped, *args):
     return map_dense(f, num_mapped, *args)
 
 
+@map_impl.py_impl(FunctionalTensorMode)
+def map_functional_tensor_mode(f, num_mapped, *args):
+    from torch._functorch.aot_autograd import from_fun, to_fun
+
+    xs = args[:num_mapped]
+    pos_args = args[num_mapped:]
+    unwrapped_xs = pytree.tree_map_only(FunctionalTensor, from_fun, xs)
+    unwrapped_args = pytree.tree_map_only(FunctionalTensor, from_fun, pos_args)
+
+    functional_map_fn = dispatch_functionalize(f)
+
+    with unset_functional_temporarily():
+        map_return = map_impl(
+            functional_map_fn, num_mapped, *unwrapped_xs, *unwrapped_args
+        )
+        return pytree.tree_map_only(torch.Tensor, to_fun, map_return)
+
```

Comment:
Could we write a generic HigherOrderOp implementation for FunctionalTensorMode and register it by default to all HigherOrderOp? It sounds like for ATen operators:
- FunctionalTensorMode unwraps the FunctionalTensor
- then does a redispatch
- the actual functionalization kernel is handled at DispatchKey::Functionalize

So to be consistent, we'd want the same for HigherOrderOp. There are two approaches we could do here:
- just have HigherOrderOp's FunctionalTensorMode implementatation call `FunctionalTensorMode.__torch_dispatch__(higherorderop, args, kwargs)`. (Might not work too well, since `__torch_dispatch__` expects OpOverload)
- Write some code that unwraps all the inputs, does the TLS-y stuff, and then does a redispatch

Reply:
Why isn't the map Functional Tensor rule just the following:
```

@map_impl.py_impl(FunctionalTensorMode)
def map_functional_tensor_mode(f, num_mapped, *args):
    unwrapped = pytree.tree_map_only(FunctionalTensor, from_fun, args)
    with unset_functional_temporarily():
        map_return = map_impl(f, num_mapped, unwrapped)
    return pytree.tree_map_only(torch.Tensor, to_fun, map_return)
```

Reply:
hmm I think you're right - I'll update it to that.

It currently looks slightly different:
```
@map_impl.py_impl(FunctionalTensorMode)
def map_functional_tensor_mode(f, num_mapped, *args):
    xs = args[:num_mapped]
    pos_args = args[num_mapped:]
    unwrapped_xs = pytree.tree_map_only(FunctionalTensor, from_fun, xs)
    unwrapped_args = pytree.tree_map_only(FunctionalTensor, from_fun, pos_args)

    functional_map_fn = dispatch_functionalize(f)

    with unset_functional_temporarily():
        map_return = map_impl(
            functional_map_fn, num_mapped, *unwrapped_xs, *unwrapped_args
        )
        return pytree.tree_map_only(torch.Tensor, to_fun, map_return)
```

I separated args into xs and pos_args to copy what the `Functionalize` kernel does - but that's probably unnecessary, since we know that the Functionalize kernel will handle it.

Reply:
If we can update the map FunctionalTensor rule to that, we can also update the cond FunctionalTensor rule to that, right? If so then we can have a single generic FunctionalTensor implementation for HigherOrderOp

Reply:
`cond` is implemented almost the same way right now, but different set of arguments:
```
def cond_functional_tensor_mode(pred, true_fn, false_fn, inputs):
    from torch._functorch.aot_autograd import from_fun, to_fun

    unwrapped_inputs = pytree.tree_map_only(FunctionalTensor, from_fun, inputs)
    unwrapped_pred = pytree.tree_map_only(FunctionalTensor, from_fun, pred)

    # We can rely on the Functionalize key to detect cond branches that modify the input
    functional_true = dispatch_functionalize(true_fn)
    functional_false = dispatch_functionalize(false_fn)

    with unset_functional_temporarily():
        cond_return = cond_op(
            unwrapped_pred, functional_true, functional_false, unwrapped_inputs
        )
    return pytree.tree_map_only(torch.Tensor, to_fun, cond_return)
```

Where we have to know that arguments 1 2 and 3 correspond to `pred` (need to unwrap) `true_fn` (need to functionalize) and `false_fn)` (need to functionalize).

It seems like a pain make a single FunctionalTensorMode function that's generic enough to work for all possible higher order op schemas. What did you have in mind? I could imagine something like this, although it seems a bit brittle since it might break if we add a new HigherOrderOp with a unique schema (was there something else you had in mind?)

```
def generic_functional_tensor_rule(
    higher_order_op: Callable,
    pre_fn_inputs: List[TensorLikeArg],
    fns: List[InnerFnArg],
    post_fn_inputs: List[TensorLikeArg]
):
    from torch._functorch.aot_autograd import from_fun, to_fun

    unwrapped_pre_args = pytree.tree_map_only(FunctionalTensor, from_fun, pre_fn_inputs)
    unwrapped_post_args = pytree.tree_map_only(FunctionalTensor, from_fun, post_fn_inputs)

    functional_fns = [dispatch_functionalize(f) for f in fns]

    with unset_functional_temporarily():
        out = higher_order_op(
            *unwrapped_pre_args, *functional_fns, *unwrapped_post_args
        )
    return pytree.tree_map_only(torch.Tensor, to_fun, out)
    
# to get the schemas to match up, we'd have to register them to map/cond like so:
def cond_functional_tensor_mode(pred, true_fn, false_fn, inputs):
    return generic_functional_tensor_rule(cond, [pred], [true_fn, false_fn], inputs)
def map_functional_tensor_mode(f, num_mapped, *args):
    return generic_functional_tensor_rule(map, [], [num_mapped], [num_mapped] + list(args))
```

Reply:
Does the following work?

```
def generic_functional_tensor_rule(
    higher_order_op,
    *args
):
    from torch._functorch.aot_autograd import from_fun, to_fun

    unwrapped_args = pytree.tree_map_only(FunctionalTensor, from_fun, args)

    with unset_functional_temporarily():
        out = higher_order_op(
            *unwrapped_args,
        )
    return pytree.tree_map_only(torch.Tensor, to_fun, out)
```

Reply:
> I don't think there's a reason to functionalize the function args in the functional_tensor_rule

Oh I see. We'll actually need to wrap the inner functions in `dispatch_functionalize` in order for functionalizing subclasses to do the right thing on higher order ops (forcing functionalization to run underneath the subclass's `__torch_dispatch__`, **inside** of the inner_function that we're tracing out.

I guess we don't need this immediately, since none of the subclass that care about compile work with higher order ops anyway as you pointed out (since e.g. DTensor doesn't implement a rule for cond and friends).

One reason I can see doing this now (other than the fact that we'll need it at some point anyway) is just because it makes my mental model of functionalization simpler: During AOTAutograd, we will always use python functionalization to trace (both the outer graph, and any inner higher order ops). What do you think?